crates/matrix-sdk-crypto/src/session_manager/group_sessions/share_strategy.rs

```diff
@@ -42,6 +46,21 @@ pub enum CollectStrategy {
         ///       trusted via interactive verification.
         ///     - It is the current own device of the user.
         only_allow_trusted_devices: bool,
+        /// If true, when a verified user has an unsigned device the key sharing
+        /// will fail with an error. If false the key will be
+        /// distributed to that unsigned device. In order to resolve
+        /// this sharing error the user can choose to ignore or blacklist the
+        /// device:
+        /// - [`Device::set_local_trust`] using [`LocalTrust::Ignored`], this
+        ///   will ignore the warning for this device, and send it the key.
+        /// - [`Device::set_local_trust`] using [`LocalTrust::BlackListed`],
+        ///   this will not distribute the key to this device.
+        ///
+        /// Once the problematic devices are blacklisted or whitelisted the
+        /// caller can retry to share a second time. This has to be done
+        /// once per problematic device.
+        #[serde(default)]
```

Comment:
is this persisted somewhere, that makes the `default` necessary?

Reply:
`EncryptionSettings` is persisted in the OutboundSession at least, and maybe some clients could persist the encryption settings? All the things we put as serializable could be persisted... (are we serializing things for other reasons? like for bindings to pass things back and forth?)
We really need to improve that part of the sdk, it's too easy to make a change that would break applications.  
I thought about having tests with json exports of all the serializable things that are persisted, that just tries to load them back and see if it breaks. This would be minimum. Or we need to think about a specific trait for things that are expected to be serialized persisted, with a version of some form

Reply:
> `EncryptionSettings` is persisted in the OutboundSession at least

Hrm... does that means that, once we enable the new behaviour, it won't work until megolm sessions are replaced?

Reply:
I don't think so. But there is something odd with the settings how they are now.

The settings persisted is used mainly (i think) in the code to detect if the session should be rotated.
I think that only part of the `EncryptionSetting` should be persisted (rotation_period, rotation_period_msgs, and history_visibility?).
The old `only_allow_trusted_devices` that was in the EncryptionSetting (replaced now by CollectStrategy) shouldn't be persisted.  Because anyhow if a change on this setting change the set of devices that should receive it, then `is_session_overshared_for_user` will detect it and rotate the session.

`history_visibility` is different because you want to rotate if you decide that this session could be shared now as part of history sharing for example, event if it doesn't change right now who get the key

Reply:
I'm not really following all of this. It sounds like we shouldn't be persisting `EncryptionSettings` at all to me.

Reply:
If you change the `rotation_period` you would want the session to be rotated immediatly instead of having to wait for the old one to be rotated based on the old rotation period. Same for `history_visibility`, you want to make history sharable from the point you change to that setting and not make early history available?

Reply:
Both of those things sound like we should be taking notice of the *current* `EncryptionSettings`, not whatever was persisted to the store.